src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt

```diff
@@ -239,10 +240,89 @@ class LocationsReconciliationService(
       if (expectedDpsAttributesSize(nomis.profiles) != (dps.attributes?.size ?: 0)) return "Cell attributes mismatch"
     }
     if (dps.residentialHousingType == null && (nomis.usages?.size ?: 0) != (dps.usage?.size ?: 0)) return "Location usage mismatch"
-    if (expectedDpsHistorySize(nomis.amendments) != (dps.changeHistory?.size ?: 0)) return "Location history mismatch"
+    if (historyDoesNotMatch(nomis, dps)) {
+      return "Location history mismatch:\n${
+        reportHistory(nomis.amendments, dps.changeHistory)
+      }"
+    }
     return null
   }
 
+  fun historyDoesNotMatch(nomis: LocationResponse, dps: Location): Boolean {
+    if (expectedDpsHistorySize(nomis.amendments) == (dps.changeHistory?.size ?: 0)) {
+      return false
+    }
+    val nomisListConverted = nomis.amendments!!.map { a ->
+      ChangeHistory(toHistoryAttribute(a.columnName), a.amendedBy, a.amendDateTime, a.oldValue, a.newValue)
+    }
+    val nomisSetSorted = nomisListConverted.sortedBy { it.amendedDate + it.attribute + it.amendedBy + it.oldValue + it.newValue }
+    val dpsSetSorted = dps.changeHistory!!.sortedBy { it.amendedDate + it.attribute + it.amendedBy + it.oldValue + it.newValue }
+
```

Comment:
minor Kotlin style things

...
You have 2 list so you could use zip to combine them then you don't need the `dpsSetSorted.elementAtOrNull(dpsIndex)`

you could of also done `forEachIndexed`

and lastly you could add a `isSame` method on either of the NOMIS or DPS objects to  then you could use something like list.any or something similar
or alternatively a extension function that returns the same data class from both objects and then use `==`

Obviously code works now but you might want to use Kotlin to improve

Reply:
Actually this comparison code was generated by the Intellij AI ! And it did originally use forEachIndexed but I needed to manipulate the list position at one point. I left all of this in partly because I might need it in future (the locations req are still somewhat influx).
I like the idea of using zip though - will try it next time